### DISCUSSION TITLE
pypi packages github.com changelogUrl not being set when overriding registryUrl

### DISCUSSION THREAD

**Discussion by sigma67** [renovatebot/renovate#25663] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Changelogs should be fetched regardless of index url

### What you tried so far.

Since we are mirroring pypi.org via Artifactory internally, we set the default index url to our internal Artifactory server. 

This also prevents useless requests to pypi.org for our internal packages which would result in 404s every time.

However, it seems this breaks the fetching of changelogs from GitHub - which would seem unrelated? I'm not sure why pypi.org is required to fetch changelogUrls for Python packages, although this may be expected behavior.

### Relevant debug logs

As you can see in the first log, the changelogUrl for isort is missing when overriding the index url. mypy's is still there, but it doesn't point to github.com

In the second log, the github.com changelogUrl is there - seemingly fetched from pypi.org?

**Log with overriding pypi index URL** 
<details><summary>Logs</summary>

``` json
               {
                 "packageName": "isort",
                 "depName": "dev/isort",
                 "datasource": "pypi",
                 "depType": "tool.pdm.dev-dependencies",
                 "currentValue": ">=5.12.0",
                 "registryUrls": [
                   "private-artifactory-simple-index-url",
                   "private-artifactory-simple-index-url"
                 ],
                 "updates": [],
                 "versioning": "pep440",
                 "warnings": [],
                 "currentVersion": "5.12.0"
               },
               {
                 "packageName": "mypy",
                 "depName": "dev/mypy",
                 "datasource": "pypi",
                 "depType": "tool.pdm.dev-dependencies",
                 "currentValue": ">=1.6.1",
                 "registryUrls": [
                   "private-artifactory-simple-index-url",
                   "private-artifactory-simple-index-url"
                 ],
                 "updates": [],
                 "versioning": "pep440",
                 "warnings": [],
                 "sourceUrl": "https://github.com/python/mypy",
                 "changelogUrl": "https://mypy-lang.blogspot.com/",
                 "currentVersion": "1.6.1"
               },
```

</details>

**Log without overriding pypi index URL**

<details><summary>Logs</summary>

``` json

               {
                 "packageName": "isort",
                 "depName": "dev/isort",
                 "datasource": "pypi",
                 "depType": "tool.pdm.dev-dependencies",
                 "currentValue": ">=5.12.0",
                 "registryUrls": [
                   "https://pypi.org/pypi/",
                   "private-artifactory-simple-index-url"
                 ],
                 "updates": [],
                 "versioning": "pep440",
                 "warnings": [],
                 "sourceUrl": "https://github.com/pycqa/isort",
                 "homepage": "https://pycqa.github.io/isort/",
                 "changelogUrl": "https://github.com/pycqa/isort/blob/main/CHANGELOG.md",
                 "currentVersion": "5.12.0"
               },
               {
                 "packageName": "mypy",
                 "depName": "dev/mypy",
                 "datasource": "pypi",
                 "depType": "tool.pdm.dev-dependencies",
                 "currentValue": ">=1.6.1",
                 "registryUrls": [
                   "https://pypi.org/pypi/",
                   "private-artifactory-simple-index-url"
                 ],
                 "updates": [],
                 "versioning": "pep440",
                 "warnings": [],
                 "sourceUrl": "https://github.com/python/mypy",
                 "homepage": "https://www.mypy-lang.org/",
                 "changelogUrl": "https://mypy-lang.blogspot.com/",
                 "currentVersion": "1.6.1"
               },

```

</details>

**Comment by rarkins** [comment]
If the URL is missing then the most likely reason is that your Artifactory is not returning complete metadata, perhaps because it supports only the "simple" pypi interface

**Reply by sigma67** [reply]
That's what I figured, Artifactory doesn't support the newer metadata interfaces (although I know nothing about how this works in renovate).

What confuses me is that the mypy changelogUrl is still there without pypi.org? The only difference being that it is not on github.com?

**Reply by rarkins** [reply]
I'm not sure why that would be. You would need to look for which exact URLs Renovate contacts, then curl them manually to look at the data and guess why it is

**Reply by sigma67** [reply, marked as answer]
I see why now. There are hardcoded changelog URLs for some projects in renovate: https://github.com/renovatebot/renovate/pull/7289/files

If the pypi JSON metadata is available, the changelog is instead pulled from the repository URL via fallback: https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/pypi/index.ts#L144

If it is not available, it falls back to the Simple API (as for artifactory right now). Then if there is no hardcoded URL, you end up without an URL.

Will have to wait until Artifactory supports the JSON API, I guess. https://jfrog.atlassian.net/browse/RTFACT-26891

**Reply by sigma67** [reply]
I managed to work around the issue by setting the private index url only for our private package prefixes.

```json

    {
      "matchDatasources": ["pypi"],
      "matchPackagePatterns": ["^INTERNALPREFIX(-|_).*" ],
      "registryUrls": [
          "private-index-url"
      ]
    },
```

**Comment by ghost** [comment]
We have the same symptoms when using the [gitlab internal pypi registry](https://docs.gitlab.com/ee/user/packages/pypi_repository/#install-from-the-group-level). The strange thing IMO here is that for all non-internal packages the registry just returns a redirect to pypi.org which should have all the information renovate needs to fetch the changelog? Renovate already follows the redirect as it detects package updates just fine. It just doesn't do changelogs anymore (though not for *all* packages, but most?)

For now I've also added the workaround from https://github.com/renovatebot/renovate/discussions/25663#discussioncomment-7521600 but this requires internal packages following a specific naming scheme.
